File: src/main/java/com/srikar/tally/controller/GroupController.java
package com.srikar.tally.controller;

import com.srikar.tally.configuration.FirebaseUserPrincipal;
import com.srikar.tally.dto.GroupRequestDto;
import com.srikar.tally.dto.GroupResponseDto;
import com.srikar.tally.service.GroupService;
import lombok.Builder;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/v1/groups")
public class GroupController {
    private final GroupService groupService;

    public GroupController(GroupService groupService) {
        this.groupService = groupService;
    }
    @GetMapping("/")
    public ResponseEntity<List<GroupResponseDto>> getAllGroupsByUser(
            @AuthenticationPrincipal FirebaseUserPrincipal userPrincipal
            ){
        var groups = groupService.getAllGroupsByUser(userPrincipal.getUid());
        return ResponseEntity.ok(groups);
    }
    @GetMapping("/{id}")
    public ResponseEntity<GroupResponseDto> getGroupByGroupId(
            @PathVariable("id") int id
    ){
        var groupFound = groupService.findGroupById(id);
        return ResponseEntity.ok(groupFound);
    }
    @PostMapping("/")
    public ResponseEntity<GroupResponseDto> createNewGroup(
            @AuthenticationPrincipal FirebaseUserPrincipal userPrincipal,
            @Validated({Builder.Default.class}) @RequestBody GroupRequestDto dto
            ){
        var groupCreated = groupService.createNewGroup(userPrincipal.getUid(),dto);
        var location = URI.create("api/v1/groups/" + groupCreated.getId());
        return ResponseEntity.created(location).body(groupCreated);
    }
    @PutMapping("/{id}")
    public ResponseEntity<GroupResponseDto> updateGroup(
            @PathVariable("id") int id,
            @Validated({Builder.Default.class}) @RequestBody GroupRequestDto dto
    ){
        var groupUpdated = groupService.updateGroup(id, dto);
        return ResponseEntity.ok(groupUpdated);
    }
    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteGroup(
            @PathVariable("id") int id
    ){
        groupService.deleteGroup(id);
        return ResponseEntity.noContent().build();
    }
    // this is where you add methods to remove members of a group.
}

File: src/main/java/com/srikar/tally/groups/controller/GroupController.kt
package com.srikar.tally.groups.controller


import com.srikar.tally.configuration.FirebaseUserPrincipal
import com.srikar.tally.groups.dto.GroupDto
import com.srikar.tally.groups.dto.NewGroupDto
import com.srikar.tally.groups.service.GroupService
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.net.URI

@RestController
@RequestMapping("/v1/groups")
class GroupController(
    private val groupService: GroupService,
) {
    @PostMapping("/")
    fun createGroup(
        @AuthenticationPrincipal user: FirebaseUserPrincipal,
        @RequestBody data: NewGroupDto
    ): ResponseEntity<Int>{
        val userId = user.uid
        val result = groupService.createGroup(userId, data)
        val location = URI.create("/api/v1/groups/")
        return ResponseEntity.created(location).body(result.get())
    }
    @GetMapping("/{groupId}")
    fun getGroupById(
        @PathVariable("groupId") groupId: Int
    ): ResponseEntity<GroupDto>{
        val group = groupService.getGroupById(groupId)
        return ResponseEntity.ok(group.get())
    }
    @GetMapping("/")
    fun getGroupByUser(
        @AuthenticationPrincipal user: FirebaseUserPrincipal
    ): ResponseEntity<List<GroupDto>>{
        val userId = user.uid
        val results = groupService.getUserGroups(userId)
        return ResponseEntity.ok(results)
    }
    @PutMapping("/{groupId}")
    fun putGroup(@AuthenticationPrincipal user: FirebaseUserPrincipal,
                 @RequestBody data: NewGroupDto,
                 @PathVariable groupId: Int): ResponseEntity<Void>{
        groupService.updateGroup(groupId, data)
        return ResponseEntity.status(204).body(null)
    }
    @DeleteMapping("/{groupId}")
    fun deleteGroup(@PathVariable groupId: Int): ResponseEntity<Void>{
        groupService.deleteGroup(groupId)
        return ResponseEntity.noContent().build();
    }
}